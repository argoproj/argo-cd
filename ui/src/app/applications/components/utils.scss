.propagation-policy-list {
    display: flex;
    justify-content: left;
}

.propagation-policy-option {
    display: flex;
    align-items: center;
    padding-right: 2em;

    label {
        color: #6D7F8B;
        font-size: 15px;
    }
}

.radio-button {
    margin-right: 0.1em;       
}