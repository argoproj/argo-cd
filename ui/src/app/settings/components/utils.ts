export function convertExpiresInToSeconds(expiresIn: string): number {
    if (!expiresIn) {
        return 0;
    }
    const time = expiresIn.match('^([0-9]+)([smhd])$');
    const duration = parseInt(time[1], 10);
    let interval = 1;
    if (time[2] === 'm') {
        interval = 60;
    } else if (time[2] === 'h') {
        interval = 60 * 60;
    } else if (time[2] === 'd') {
        interval = 60 * 60 * 24;
    }
    return duration * interval;
}

export function validExpiresIn(expiresIn: string): boolean {
    if (!expiresIn) {
        return true;
    }
    return expiresIn.match('^([0-9]+)([smhd])$') !== null;
}
