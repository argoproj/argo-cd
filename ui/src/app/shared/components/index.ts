export {DataLoader, ErrorNotification} from 'argo-ui';

export * from './array-input/array-input';
export * from './expandable/expandable';
export * from './page';
export * from './checkbox/checkbox-field';
export * from './colors';
export * from './cluster';
export * from './connection-state-icon';
export * from './revision-help-icon';
export * from './query';
export * from './events-list/events-list';
export * from './editable-panel/editable-panel';
export * from './tags-input/tags-input-field';
export * from './paginate/paginate';
export * from './empty-state/empty-state';
export * from './tags-input/tags-input';
export * from './monaco-editor';
export * from './yaml-editor/yaml-editor';
export * from './progress/progress-popup';
export * from './info-popup/info-popup';
export * from './repo';
export * from './revision';
export * from './timestamp';
export * from './spinner';
export * from './badge-panel/badge-panel';
export * from './number-field';
