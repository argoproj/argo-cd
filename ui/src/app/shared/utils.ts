export function hashCode(str: string) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        // tslint:disable-next-line:no-bitwise
        hash = ~~((hash << 5) - hash + str.charCodeAt(i));
    }
    return hash;
}
