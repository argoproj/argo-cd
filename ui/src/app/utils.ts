export const getBase = () => {
    const bases = document.getElementsByTagName('base');

    return bases.length > 0 ? bases[0].getAttribute('href') || '/' : '/';
};
