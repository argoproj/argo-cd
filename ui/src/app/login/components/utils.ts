import {
    AuthorizationServer,
    Client,
    authorizationCodeGrantRequest,
    calculatePKCECodeChallenge,
    discoveryRequest,
    expectNoState,
    generateRandomCodeVerifier,
    isOAuth2Error,
    parseWwwAuthenticateChallenges,
    processAuthorizationCodeOpenIDResponse,
    processDiscoveryResponse,
    validateAuthResponse
} from 'oauth4webapi';
import {AuthSettings} from '../../shared/models';

export const discoverAuthServer = (issuerURL: URL): Promise<AuthorizationServer> => discoveryRequest(issuerURL).then(res => processDiscoveryResponse(issuerURL, res));

export const PKCECodeVerifier = {
    get: () => sessionStorage.getItem(window.btoa('code_verifier')),
    set: (codeVerifier: string) => sessionStorage.setItem(window.btoa('code_verifier'), codeVerifier),
    unset: () => sessionStorage.removeItem(window.btoa('code_verifier'))
};

export const getPKCERedirectURI = () => {
    const currentOrigin = new URL(window.location.origin);

    currentOrigin.pathname = '/pkce/verify';

    return currentOrigin;
};

export class PKCELoginError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'PKCELoginError';
    }
}

const validateAndGetOIDCForPKCE = async (oidcConfig: AuthSettings['oidcConfig']) => {
    if (!oidcConfig) {
        throw new PKCELoginError('No OIDC Config found');
    }

    let issuerURL: URL;
    try {
        issuerURL = new URL(oidcConfig.issuer);
    } catch (e) {
        throw new PKCELoginError(`Invalid oidc issuer ${oidcConfig.issuer}`);
    }

    if (!oidcConfig.clientID) {
        throw new PKCELoginError('No OIDC Client Id found');
    }

    let authorizationServer: AuthorizationServer;
    try {
        authorizationServer = await discoverAuthServer(issuerURL);
    } catch (e) {
        throw new PKCELoginError(e);
    }

    return {
        issuerURL,
        authorizationServer,
        clientID: oidcConfig.clientID
    };
};

export const pkceLogin = async (oidcConfig: AuthSettings['oidcConfig'], redirectURI: string) => {
    const {authorizationServer} = await validateAndGetOIDCForPKCE(oidcConfig);

    if (!authorizationServer.authorization_endpoint) {
        throw new PKCELoginError('No Authorization Server endpoint found');
    }

    if (!authorizationServer?.code_challenge_methods_supported?.includes('S256')) {
        throw new PKCELoginError('Authorization Server does not support S256 code challenge method');
    }

    const codeVerifier = generateRandomCodeVerifier();

    const codeChallange = await calculatePKCECodeChallenge(codeVerifier);

    const authorizationServerConsentScreen = new URL(authorizationServer.authorization_endpoint);

    authorizationServerConsentScreen.searchParams.set('client_id', oidcConfig.clientID);
    authorizationServerConsentScreen.searchParams.set('code_challenge', codeChallange);
    authorizationServerConsentScreen.searchParams.set('code_challenge_method', 'S256');
    authorizationServerConsentScreen.searchParams.set('redirect_uri', redirectURI);
    authorizationServerConsentScreen.searchParams.set('response_type', 'code');
    authorizationServerConsentScreen.searchParams.set('scope', oidcConfig.scopes.join(' '));

    PKCECodeVerifier.set(codeVerifier);

    window.location.replace(authorizationServerConsentScreen.toString());
};

export const pkceCallback = async (queryParams: string, oidcConfig: AuthSettings['oidcConfig'], redirectURI: string) => {
    const codeVerifier = PKCECodeVerifier.get();

    if (!codeVerifier) {
        throw new PKCELoginError('No code verifier found in session');
    }

    let callbackQueryParams = new URLSearchParams();
    try {
        callbackQueryParams = new URLSearchParams(queryParams);
    } catch (e) {
        throw new PKCELoginError('Invalid query parameters');
    }

    if (!callbackQueryParams.get('code')) {
        throw new PKCELoginError('No code in query parameters');
    }

    if (callbackQueryParams.get('state') === '') {
        callbackQueryParams.delete('state');
    }

    const {authorizationServer} = await validateAndGetOIDCForPKCE(oidcConfig);

    const client: Client = {
        client_id: oidcConfig.clientID,
        token_endpoint_auth_method: 'none'
    };

    const params = validateAuthResponse(authorizationServer, client, callbackQueryParams, expectNoState);

    if (isOAuth2Error(params)) {
        throw new PKCELoginError('Error validating auth response');
    }

    const response = await authorizationCodeGrantRequest(authorizationServer, client, params, redirectURI, codeVerifier);

    const authChallengeExtract = parseWwwAuthenticateChallenges(response);

    if (authChallengeExtract?.length > 0) {
        throw new PKCELoginError('Error parsing authentication challenge');
    }

    const result = await processAuthorizationCodeOpenIDResponse(authorizationServer, client, response);

    if (isOAuth2Error(result)) {
        throw new PKCELoginError(`Error getting token ${result.error_description}`);
    }

    if (!result.id_token) {
        throw new PKCELoginError('No token in response');
    }

    document.cookie = `argocd.token=${result.id_token}; path=/`;

    window.location.replace('/applications');
};
