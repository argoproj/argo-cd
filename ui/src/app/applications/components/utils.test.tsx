import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Application, HealthStatus, HealthStatuses, OperationPhases, ResourceResult, ResultCodes, SyncStatuses} from '../../shared/models';
import {ComparisonStatusIcon, getAppOperationState, getOperationType, HealthStatusIcon, OperationState, ResourceResultIcon} from './utils';

const zero = new Date(0).toISOString();

test('getAppOperationState.DeletionTimestamp', () => {
    const state = getAppOperationState({metadata: {deletionTimestamp: zero}} as Application);

    expect(state).toStrictEqual({phase: OperationPhases.Running, startedAt: zero});
});

test('getAppOperationState.Operation', () => {
    const state = getAppOperationState({metadata: {}, operation: {}} as Application);

    expect(state.phase).toBe(OperationPhases.Running);
    expect(state.startedAt).toBeDefined();
    expect(state.operation).toStrictEqual({sync: {}});
});

test('getAppOperationState.Status', () => {
    const state = getAppOperationState({
        metadata: {},
        status: {operationState: {phase: OperationPhases.Error, startedAt: zero}}
    } as Application);

    expect(state.phase).toBe(OperationPhases.Error);
});

test('getOperationType.Delete', () => {
    const state = getOperationType({metadata: {deletionTimestamp: zero.toString()}} as Application);

    expect(state).toBe('Delete');
});

test('getOperationType.Sync.Operation', () => {
    const state = getOperationType({metadata: {}, operation: {sync: {}}} as Application);

    expect(state).toBe('Sync');
});

test('getOperationType.DeleteAndRecentSync', () => {
    const state = getOperationType({metadata: {deletionTimestamp: '123'}, status: {operationState: {operation: {sync: {}}}}} as Application);

    expect(state).toBe('Delete');
});

test('getOperationType.Sync.Status', () => {
    const state = getOperationType({metadata: {}, status: {operationState: {operation: {sync: {}}}}} as Application);

    expect(state).toBe('Sync');
});

test('getOperationType.Unknown', () => {
    const state = getOperationType({metadata: {}, status: {}} as Application);

    expect(state).toBe('Unknown');
});

test('OperationState.undefined', () => {
    const tree = renderer.create(<OperationState app={{metadata: {}, status: {}} as Application} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.quiet', () => {
    const tree = renderer.create(<OperationState app={{metadata: {}, status: {operationState: {}}} as Application} quiet={true} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Unknown', () => {
    const tree = renderer.create(<OperationState app={{metadata: {}, status: {operationState: {}}} as Application} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Deleting', () => {
    const tree = renderer.create(<OperationState app={{metadata: {deletionTimestamp: zero}, status: {operationState: {}}} as Application} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Sync OK', () => {
    const tree = renderer
        .create(<OperationState app={{metadata: {}, status: {operationState: {operation: {sync: {}}, phase: OperationPhases.Succeeded}}} as Application} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Sync error', () => {
    const tree = renderer.create(<OperationState app={{metadata: {}, status: {operationState: {operation: {sync: {}}, phase: OperationPhases.Error}}} as Application} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Sync failed', () => {
    const tree = renderer.create(<OperationState app={{metadata: {}, status: {operationState: {operation: {sync: {}}, phase: OperationPhases.Failed}}} as Application} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('OperationState.Syncing', () => {
    const tree = renderer
        .create(<OperationState app={{metadata: {}, status: {operationState: {operation: {sync: {}}, phase: OperationPhases.Running}}} as Application} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('ComparisonStatusIcon.Synced', () => {
    const tree = renderer.create(<ComparisonStatusIcon status={SyncStatuses.Synced} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ComparisonStatusIcon.OutOfSync', () => {
    const tree = renderer.create(<ComparisonStatusIcon status={SyncStatuses.OutOfSync} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ComparisonStatusIcon.Unknown', () => {
    const tree = renderer.create(<ComparisonStatusIcon status={SyncStatuses.Unknown} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('HealthStatusIcon.Unknown', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Unknown} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});
test('HealthStatusIcon.Progressing', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Progressing} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('HealthStatusIcon.Suspended', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Suspended} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('HealthStatusIcon.Healthy', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Healthy} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('HealthStatusIcon.Degraded', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Degraded} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});
test('HealthStatusIcon.Missing', () => {
    const tree = renderer.create(<HealthStatusIcon state={{status: HealthStatuses.Missing} as HealthStatus} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Synced', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{status: ResultCodes.Synced, message: 'my-message'} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Pruned', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{status: ResultCodes.Pruned} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.SyncFailed', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{status: ResultCodes.SyncFailed} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Hook.Running', () => {
    const tree = renderer
        .create(
            <ResourceResultIcon
                resource={
                    {
                        hookType: 'Sync',
                        hookPhase: OperationPhases.Running,
                        message: 'my-message'
                    } as ResourceResult
                }
            />
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Hook.Failed', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{hookType: 'Sync', hookPhase: OperationPhases.Failed} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Hook.Error', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{hookType: 'Sync', hookPhase: OperationPhases.Error} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Hook.Succeeded', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{hookType: 'Sync', hookPhase: OperationPhases.Succeeded} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('ResourceResultIcon.Hook.Terminating', () => {
    const tree = renderer.create(<ResourceResultIcon resource={{hookType: 'Sync', hookPhase: OperationPhases.Terminating} as ResourceResult} />).toJSON();

    expect(tree).toMatchSnapshot();
});
