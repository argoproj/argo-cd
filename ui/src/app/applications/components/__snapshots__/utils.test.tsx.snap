// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComparisonStatusIcon.OutOfSync 1`] = `
Array [
  <i
    className="fa fa-arrow-alt-circle-up"
    qe-id="utils-sync-status-title"
    style={
      Object {
        "color": "#f4c030",
      }
    }
    title="OutOfSync"
  />,
  " ",
]
`;

exports[`ComparisonStatusIcon.Synced 1`] = `
Array [
  <i
    className="fa fa-check-circle"
    qe-id="utils-sync-status-title"
    style={
      Object {
        "color": "#18BE94",
      }
    }
    title="Synced"
  />,
  " ",
]
`;

exports[`ComparisonStatusIcon.Unknown 1`] = `
Array [
  <i
    className="fa fa-circle-notch fa-spin"
    qe-id="utils-sync-status-title"
    style={
      Object {
        "color": "#CCD6DD",
      }
    }
    title="Unknown"
  />,
  " ",
]
`;

exports[`HealthStatusIcon.Degraded 1`] = `
<i
  className="fa fa-heart-broken"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#E96D76",
    }
  }
  title="Degraded"
/>
`;

exports[`HealthStatusIcon.Healthy 1`] = `
<i
  className="fa fa-heart"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#18BE94",
    }
  }
  title="Healthy"
/>
`;

exports[`HealthStatusIcon.Missing 1`] = `
<i
  className="fa fa-ghost"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#CCD6DD",
    }
  }
  title="Missing"
/>
`;

exports[`HealthStatusIcon.Progressing 1`] = `
<i
  className="fa fa fa-circle-notch fa-spin"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#0DADEA",
    }
  }
  title="Progressing"
/>
`;

exports[`HealthStatusIcon.Suspended 1`] = `
<i
  className="fa fa-heart"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#CCD6DD",
    }
  }
  title="Suspended"
/>
`;

exports[`HealthStatusIcon.Unknown 1`] = `
<i
  className="fa fa-ghost"
  qe-id="utils-health-status-title"
  style={
    Object {
      "color": "#CCD6DD",
    }
  }
  title="Unknown"
/>
`;

exports[`OperationState.Deleting 1`] = `
Array [
  <i
    className="fa fa-circle-notch fa-spin"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#0DADEA",
      }
    }
    title="Deleting"
  />,
  " ",
  "Deleting",
]
`;

exports[`OperationState.Sync OK 1`] = `
Array [
  <i
    className="fa fa-check-circle"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#18BE94",
      }
    }
    title="Sync OK"
  />,
  " ",
  "Sync OK",
]
`;

exports[`OperationState.Sync error 1`] = `
Array [
  <i
    className="fa fa-times-circle"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#E96D76",
      }
    }
    title="Sync error"
  />,
  " ",
  "Sync error",
]
`;

exports[`OperationState.Sync failed 1`] = `
Array [
  <i
    className="fa fa-times-circle"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#E96D76",
      }
    }
    title="Sync failed"
  />,
  " ",
  "Sync failed",
]
`;

exports[`OperationState.Syncing 1`] = `
Array [
  <i
    className="fa fa-circle-notch fa-spin"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#0DADEA",
      }
    }
    title="Syncing"
  />,
  " ",
  "Syncing",
]
`;

exports[`OperationState.Unknown 1`] = `
Array [
  <i
    className="fa fa-circle-notch fa-spin"
    qe-id="utils-operations-status-title"
    style={
      Object {
        "color": "#0DADEA",
      }
    }
    title="Unknown"
  />,
  " ",
  "Unknown",
]
`;

exports[`OperationState.quiet 1`] = `null`;

exports[`OperationState.undefined 1`] = `null`;

exports[`ResourceResultIcon.Hook.Error 1`] = `
<i
  className="fa fa-heart-broken"
  style={
    Object {
      "color": "#E96D76",
    }
  }
/>
`;

exports[`ResourceResultIcon.Hook.Failed 1`] = `
<i
  className="fa fa-heart-broken"
  style={
    Object {
      "color": "#E96D76",
    }
  }
/>
`;

exports[`ResourceResultIcon.Hook.Running 1`] = `
<i
  className="fa fa-circle-notch fa-spin"
  style={
    Object {
      "color": "#0DADEA",
    }
  }
  title="Running: my-message;"
/>
`;

exports[`ResourceResultIcon.Hook.Succeeded 1`] = `
<i
  className="fa fa-heart"
  style={
    Object {
      "color": "#18BE94",
    }
  }
/>
`;

exports[`ResourceResultIcon.Hook.Terminating 1`] = `
<i
  className="fa fa-circle-notch fa-spin"
  style={
    Object {
      "color": "#0DADEA",
    }
  }
/>
`;

exports[`ResourceResultIcon.Pruned 1`] = `
<i
  className="fa fa-heart"
  style={
    Object {
      "color": "#CCD6DD",
    }
  }
/>
`;

exports[`ResourceResultIcon.SyncFailed 1`] = `
<i
  className="fa fa-heart-broken"
  style={
    Object {
      "color": "#E96D76",
    }
  }
/>
`;

exports[`ResourceResultIcon.Synced 1`] = `
<i
  className="fa fa-heart"
  style={
    Object {
      "color": "#18BE94",
    }
  }
  title="Synced: my-message"
/>
`;
